### PR TITLE
Fix crash in sighelp

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/72012
Fixes https://github.com/dotnet/roslyn/issues/74500
Fixes https://github.com/dotnet/roslyn/issues/74383

### CODE REVIEW

File: src/EditorFeatures/Core.Wpf/SignatureHelp/Controller.Session_ComputeModel.cs [hunk 1]
```diff
@@ -101,8 +101,8 @@ internal partial class Session
                             currentModel.Provider == provider &&
                             currentModel.GetCurrentSpanInSubjectBuffer(disconnectedBufferGraph.SubjectBufferSnapshot).Span.Start == items.ApplicableSpan.Start &&
                             currentModel.Items.IndexOf(currentModel.SelectedItem) == items.SelectedItemIndex &&
-                            currentModel.ArgumentIndex == items.ArgumentIndex &&
-                            currentModel.ArgumentCount == items.ArgumentCount &&
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/SignatureHelp/Controller.Session_ComputeModel.cs:105]: renamed these to make the distinction clear.  'argument count' refferred to how many explicit argument syntax nodes are present in the code.  'argument index' was very badly named.  It indicated 'this is the *parameter* in the actual symcol we think the user is currently on'

File: src/EditorFeatures/Core.Wpf/SignatureHelp/Controller.Session_ComputeModel.cs [hunk 2]
```diff
@@ -112,14 +112,28 @@ internal partial class Session
...
+                            disconnectedBufferGraph,
+                            items.ApplicableSpan,
+                            provider,
+                            items.Items,
+                            selectedItem,
+                            items.SemanticParameterIndex,
+                            items.SyntacticArgumentCount,
+                            items.ArgumentName,
+                            selectedParameter: 0,
+                            userSelected);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/SignatureHelp/Controller.Session_ComputeModel.cs:125]: just renames + wrapping.  no other change.

File: src/EditorFeatures/Core.Wpf/SignatureHelp/Controller.Session_ComputeModel.cs [hunk 3]
```diff
@@ -112,14 +112,28 @@ internal partial class Session
...
-                        var selection = DefaultSignatureHelpSelector.GetSelection(model.Items,
-                            model.SelectedItem, model.UserSelected, model.ArgumentIndex, model.ArgumentCount, model.ArgumentName, isCaseSensitive);
+                        var selection = DefaultSignatureHelpSelector.GetSelection(
+                            model.Items,
+                            model.SelectedItem,
+                            model.UserSelected,
+                            model.SemanticParameterIndex,
+                            model.SyntacticArgumentCount,
+                            model.ArgumentName,
+                            isCaseSensitive);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/SignatureHelp/Controller.Session_ComputeModel.cs:136]: just renames + wrapping.  no other change.

File: src/EditorFeatures/Core.Wpf/SignatureHelp/Controller.Session_UpdateModel.cs [hunk 4]
```diff
@@ -67,12 +67,18 @@ private static int GetSelectedParameter(SignatureHelpItem bestItem, int paramete
...
-                private static void SelectBestItem(ref SignatureHelpItem currentItem, ref bool userSelected,
-                    IList<SignatureHelpItem> filteredItems, int selectedParameter, int argumentCount, string name, bool isCaseSensitive)
+                private static void SelectBestItem(
+                    ref SignatureHelpItem currentItem,
+                    ref bool userSelected,
+                    IList<SignatureHelpItem> filteredItems,
+                    int semanticParameterIndex,
+                    int syntacticArgumentCount,
+                    string name,
+                    bool isCaseSensitive)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/SignatureHelp/Controller.Session_UpdateModel.cs:77]: just renames + wrapping.  no other change.

File: src/EditorFeatures/Core.Wpf/SignatureHelp/Model.cs [hunk 5]
```diff
@@ -23,8 +23,12 @@ internal class Model
...
         /// <summary>UserSelected is true if the SelectedItem is the result of a user selection (up/down arrows).</summary>
         public bool UserSelected { get; }
-        public int ArgumentIndex { get; }
-        public int ArgumentCount { get; }
+
+        /// <inheritdoc cref="SignatureHelpItems.SemanticParameterIndex"/>
+        public int SemanticParameterIndex { get; }
+
+        /// <inheritdoc cref="SignatureHelpItems.SyntacticArgumentCount"/>
+        public int SyntacticArgumentCount { get; }
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/SignatureHelp/Model.cs:31]: just renames.

File: src/EditorFeatures/Core.Wpf/SignatureHelp/Controller.Session_ComputeModel.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/SignatureHelp/Controller.Session_ComputeModel.cs:105]: note: the pr is almost entirely this rename.  i call out anything that isn't just that.

File: src/Features/Core/Portable/SignatureHelp/SignatureHelpItems.cs [hunk 6]
```diff
@@ -57,24 +61,17 @@ internal class SignatureHelpItems
...
 
-        if (argumentIndex < 0)
-        {
-            throw new ArgumentException($"{nameof(argumentIndex)} < 0. {argumentIndex} < 0", nameof(argumentIndex));
-        }
-
-        if (argumentCount < argumentIndex)
-        {
-            throw new ArgumentException($"{nameof(argumentCount)} < {nameof(argumentIndex)}. {argumentCount} < {argumentIndex}", nameof(argumentIndex));
-        }
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/SignatureHelp/SignatureHelpItems.cs:77]: this was the crux of hte bug.  a check that literally made no sense.  the number of syntactic arguments provided has no basis on which semantic *parameter* we think is teh right one to select.  It's just information about the syntax tree, that's all.

**Review by dibarbet — APPROVED** [review]

**Comment by KirillOsenkov** [comment]
Thanks for fixing, this was for me the #1 most annoying/visible Roslyn bug for my usage in the past year or maybe more.

**Pull request merged**